### PR TITLE
Fix sharing resolution context for bean providers

### LINKED ISSUE #7581: Concurrent injection regression in 3.5.x
### Expected Behavior

When injecting prototype beans using a jakarta.inject.Provider injection fails in Micronaut 3.5.1 while this succeeds in 3.4.4. I suspect it has something to do with to the use of qualifiers or with generics, when I replace the Provider with a concrete class it seems to work.

### Actual Behaviour

Injection (or actually the resolving of the provider) fails with various errors:

```
java.util.concurrent.ExecutionException: io.micronaut.context.exceptions.CircularDependencyException: Failed to inject value for parameter [counter] of class: injection.bug.Injectable

Message: Circular dependency detected
Path Taken: 
Injectable.createFirst(Provider counter)
^                                     |
|                                     |
|                                     |
+-------------------------------------+
	at java.base/java.util.concurrent.ForkJoinTask.reportExecutionException(ForkJoinTask.java:605)
	at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:981)
	at injection.bug.InjectionBugTest.lambda$testConcurrency$2(InjectionBugTest.java:63)
	at java.base/java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1602)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.MatchOps$MatchTask.doLeaf(MatchOps.java:306)
	at java.base/java.util.stream.MatchOps$MatchTask.doLeaf(MatchOps.java:277)
	at java.base/java.util.stream.AbstractShortCircuitTask.compute(AbstractShortCircuitTask.java:115)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: io.micronaut.context.exceptions.CircularDependencyException: Failed to inject value for parameter [counter] of class: injection.bug.Injectable

Message: Circular dependency detected
Path Taken: 
Injectable.createFirst(Provider counter)
^                                     |
|                                     |
|                                     |
+-------------------------------------+
	at io.micronaut.context.AbstractBeanResolutionContext$DefaultPath.pushConstructorResolve(AbstractBeanResolutionContext.java:346)
	at io.micronaut.context.AbstractBeanResolutionContext$DefaultPath.pushConstructorResolve(AbstractBeanResolutionContext.java:332)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1284)
	at injection.bug.$Config$CreateFirst3$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2331)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2282)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2228)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:2989)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2879)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2773)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1594)
	at io.micronaut.inject.provider.JakartaProviderBeanDefinition.lambda$buildProvider$0(JakartaProviderBeanDefinition.java:65)
	at injection.bug.Injectee.getValue(Injectee.java:38)
	at injection.bug.Injectee.showWhatIsInjected(Injectee.java:27)
	at injection.bug.InjectionBugTest.lambda$testConcurrency$0(InjectionBugTest.java:55)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedInterruptibleCallable.exec(ForkJoinTask.java:1461)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinTask.awaitDone(ForkJoinTask.java:436)
	at java.base/java.util.concurrent.ForkJoinTask.awaitPoolInvoke(ForkJoinTask.java:1034)
	at java.base/java.util.concurrent.ForkJoinPool.invokeAll(ForkJoinPool.java:2731)
	at injection.bug.InjectionBugTest.testConcurrency(InjectionBugTest.java:42)
...
```

```
Message: Expected only one bean dependent!
Path Taken: Injectable.createFirst(Provider counter)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2344)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2282)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2228)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:2989)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2879)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2773)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1594)
	at io.micronaut.inject.provider.JakartaProviderBeanDefinition.lambda$buildProvider$0(JakartaProviderBeanDefinition.java:65)
	at injection.bug.Injectee.getValue(Injectee.java:38)
	at injection.bug.Injectee.showWhatIsInjected(Injectee.java:29)
	at injection.bug.InjectionBugTest.lambda$testConcurrency$0(InjectionBugTest.java:55)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedInterruptibleCallable.exec(ForkJoinTask.java:1461)
	... 5 more
Caused by: java.lang.IllegalStateException: Expected only one bean dependent!
	at io.micronaut.context.AbstractBeanResolutionContext.markDependentAsFactory(AbstractBeanResolutionContext.java:152)
	at injection.bug.$Config$CreateSecond4$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2331)
	... 16 more
```

```
java.util.concurrent.ExecutionException: java.lang.NullPointerException: Cannot invoke "io.micronaut.context.BeanResolutionContext$Segment.getDeclaringType()" because "segment" is null
	at java.base/java.util.concurrent.ForkJoinTask.reportExecutionException(ForkJoinTask.java:605)
	at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:981)
	at injection.bug.InjectionBugTest.lambda$testConcurrency$2(InjectionBugTest.java:63)
	at java.base/java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1602)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.MatchOps$MatchTask.doLeaf(MatchOps.java:306)
	at java.base/java.util.stream.MatchOps$MatchTask.doLeaf(MatchOps.java:277)
	at java.base/java.util.stream.AbstractShortCircuitTask.compute(AbstractShortCircuitTask.java:115)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: java.lang.NullPointerException: Cannot invoke "io.micronaut.context.BeanResolutionContext$Segment.getDeclaringType()" because "segment" is null
	at io.micronaut.context.exceptions.MessageUtils.buildMessage(MessageUtils.java:45)
	at io.micronaut.context.exceptions.BeanInstantiationException.<init>(BeanInstantiationException.java:49)
	at io.micronaut.context.DefaultBeanContext.resolveByBeanFactory(DefaultBeanContext.java:2344)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2282)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2228)
	at io.micronaut.context.DefaultBeanContext.createRegistration(DefaultBeanContext.java:2989)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2879)
	at io.micronaut.context.DefaultBeanContext.resolveBeanRegistration(DefaultBeanContext.java:2773)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1594)
	at io.micronaut.inject.provider.JakartaProviderBeanDefinition.lambda$buildProvider$0(JakartaProviderBeanDefinition.java:65)
	at injection.bug.Injectee.getValue(Injectee.java:38)
	at injection.bug.Injectee.showWhatIsInjected(Injectee.java:27)
	at injection.bug.InjectionBugTest.lambda$testConcurrency$0(InjectionBugTest.java:55)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedInterruptibleCallable.exec(ForkJoinTask.java:1461)
	... 5 more
```

### Steps To Reproduce

```
git clone 'git@github.com:auke-/micronaut-concurrent-injection-bug.git'
cd micronaut-concurrent-injection-bug
./mvnw clean test
```

### Environment Information

MacOS Monteray
OpenJDK 17

### Example Application

See https://github.com/auke-/micronaut-concurrent-injection-bug

### Version

3.5.1

### PR DESCRIPTION
**Pull request by dstepanov**
Fixes #7581

### CODE REVIEW

**Review by graemerocher — APPROVED** [review]

File: inject/src/main/java/io/micronaut/inject/provider/JakartaProviderBeanDefinition.java [hunk 1]
```diff
@@ -50,20 +50,18 @@ public boolean isPresent() {
     @Override
     protected Provider<Object> buildProvider(BeanResolutionContext resolutionContext, BeanContext context, Argument<Object> argument, Qualifier<Object> qualifier, boolean singleton) {
```
Comment (auke-) [inline inject/src/main/java/io/micronaut/inject/provider/JakartaProviderBeanDefinition.java:51]: There are 3 implementations of the `buildProvider` method: JakartaProviderBeanDefinition (this class), JavaxProviderBeanDefinition and BeanProviderDefinition. Their implementation looks very similar, shouldn't JavaxProviderBeanDefinition and BeanProviderDefinition also be changed?

Comment (graemerocher) [reply inject/src/main/java/io/micronaut/inject/provider/JakartaProviderBeanDefinition.java:51]: Good catch @dstepanov we will need to alter the other impls

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: inject/src/main/java/io/micronaut/inject/provider/JakartaProviderBeanDefinition.java (hunk 1 above)
Comment (dstepanov) [reply inject/src/main/java/io/micronaut/inject/provider/JakartaProviderBeanDefinition.java:51]: Yeah, good catch, pushed fixes for other providers as well

**Comment by dstepanov** [comment]
I have removed the `javax` test because it requires adding `javax` dependency and that alters other tests.

**Review by sdelamo — APPROVED** [review]

**Pull request merged**